keyservice/server.go

```diff
@@ -31,19 +52,19 @@ func (ks *Server) encryptWithPgp(key *PgpKey, plaintext []byte) ([]byte, error)
 func (ks *Server) encryptWithKms(key *KmsKey, plaintext []byte) ([]byte, error) {
 	ctx := make(map[string]*string)
 	for k, v := range key.Context {
-		ctx[k] = &v
+		value := v // Allocate a new string to prevent the pointer below from referring to only the last iteration value
+		ctx[k] = &value
```

Comment:
Could do `ctx[k] = &key.Context[k]` too to avoid this as well, but what you did works just fine.

Reply:
I also wonder if there's any linters that catch or warn against this, since I know for a fact this has triggered a bug at least one other time in sops.

Reply:
It would make sense to try to use `&key.Context[k]` to get pointers to the map values, but that doesn't work, because apparently Go [map values are not addressable](https://github.com/golang/go/issues/11865). So the simplest fix I could find was to just assign the value to a new variable causing a new string to be allocated.

Testing that the encryption context is correctly passed all the way to the KMS SDK would be a good idea and I was initially considering that approach for the test cases. I didn't go with that approach, because I was a little reluctant exposing the `kmsSvc kmsiface.KMSAPI` global variable to the `keyservice` package, which I thought to be the best location for these tests. If it's okay to reuse that global variable to also enable KMS API mocking in `keyservice` package, I don't see why I couldn't change the implementation to use that approach and remove the provider indirection I introduced.

Finally, using a linter to catch these types of bugs would be awesome. I'm not knowledgeable enough about the Go ecosystem to suggest a linter that would detect this problem. I did do some quick googling, but all I was able to find was [this](https://github.com/evanj/loopcheck).

Reply:
I've now pushed changes that use the already existing KMS mock. These changes have the aforementioned downside that they expose `kmsiface.KMSAPI` variable in `kms` package in order to make it accessible from `keyservice` package. It would be simpler to have this new encryption context test located within `kms` package to allow it to access that `kmsiface.KMSAPI` mock without publicly exporting the variable, but this would lead to a circular dependency between the packages (`kms` -> `keyservice` -> `kms`). 

In my opinion, it would be preferable to consider getting rid of the use of a global variable for mock injection purposes and instead refactor the `kms` package to allow tests to provide the `kmsiface.KMSAPI` implementation directly to an instance of `kms.MasterKey`. This would remove the need to make the variable public for mocking purposes. Additionally, I'm wondering if the global variable might cause any kind of concurrency issues between tests. I'm not that well-versed in Go to know how concurrency might influence these testing scenarios, but I can see that if two tests were running concurrently and attempted to modify the global `kmsiface.KMSAPI`, there would be trouble.

Reply:
Hmm, that is somewhat ugly too. I think we're taking the wrong approach to testing this. This bug was caused by the conversion from a `keyservice.KmsKey` to a `kms.MasterKey` being buggy. We should split that out into a function and test it. I feel like the kind of testing you're doing right now is better suited for our "functional tests" (end-to-end tests). Unfortunately we don't have the infrastructure right now to use KMS keys during tests, but I feel like setting that up and running integration tests would be the way forward. Testing Go code across packages can get very messy, and mocking in Go is not great. I'm not really sure the tests we currently have in the `kms` package (before this PR) are actually very useful. It'd be much better to have a real end-to-end test instead of using a mock.

Reply:
Making the global variable public is indeed ugly. On the other hand, your suggestion of extracting a helper method is a sound choice. I considered that initially, but opted to go with a mock-based test, because by using a mock I could check that the actual call path would provide the wanted encryption context. In contrast calling the helper method directly in a test will not guarantee that the actual call path will actually use that helper function, even if the test case shows the helper function to be working fine. That said, I'm fine with your suggestion, especially if a more comprehensive functional test is added in the future. I pushed new code changes that extract a helper function and test that function directly.

Reply:
Thanks for making these changes again. It looks good to me now. If my information is still correct, @ajvb will work on setting up integration tests and an upgrade path for this PR, at which point, it can be merged.

Reply:
No problem! I rebased the changes, dropping irrelevant commits.